wire-schema/src/jvmMain/java/com/squareup/wire/schema/EmittingRules.kt

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.schema
+
+/**
+ * A set of rules that describes which types to generate.
+ *
+ * Despite the builder, instances of this class are not safe for concurrent use.
+ *
+ * ### Identifier Matching
+ *
+ * Identifiers in this set may be in the following forms:
+ *
+ *  * Package names, followed by `.*`, like `squareup.protos.person.*`. This matches types and
+ *    services defined in the package and its descendant packages.
+ *
+ *  * Fully qualified type and service names, like `squareup.protos.person.Person`.
+ *
+ * Identifiers should not contain member names.
+ *
+ * This set has *included identifiers* and *excluded identifiers*, with excludes taking precedence
```

Comment:
Is this in line with the idea of honouring the most granular rule, regardless of whether it's an include or exclude? In that case I expect us to fail if a type is both included and excluded explicitly at the same time.

Reply:
Yeah it's not doing the most precise rule yet